### PR TITLE
compiler/next add AST for enums and parse them

### PR DESCRIPTION
**Pull request by mppf**
Enums are represented by:
 * EnumDecl, which declares
 * Enum, a TypeSymbol containing the EnumElementDecls
 * EnumElementDecl which declares
 * EnumElement

So for example in

``` chapel
enum myEnum {
 a, b ,c
}
```

`myEnum` is represented by an `Enum` (which is declared in an EnumDecl) 
and `a`, `b`, and `c` are `EnumElements` (which are in turn declared with
EnumElementDecls within the Enum).

This PR also adds an iterator that can ignore Comments so that it's easy 
to enumerate the EnumElementDecls or the Decls in a MultiDecl.

### CODE REVIEW

File: compiler/next/include/chpl/uast/Enum.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  /**
+   Return the number of EnumElements and Comments contained in this Enum.
+   */
+  int numStmts() const {
+    return this->numChildren();
+  }
+  /**
+   Return the i'th EnumElement or Comment in this Enum.
+   */
+  const Expression* stmt(int i) const {
```
Comment (mppf) [inline compiler/next/include/chpl/uast/Enum.h:76]: See also https://github.com/chapel-lang/chapel/pull/17680#discussion_r631532447 - maybe this should be name declOrComment. In both cases, I'm hoping to add an iterator just for the declared Syms ignoring comments.

Comment (mppf) [reply compiler/next/include/chpl/uast/Enum.h]: I renamed it to that.

File: compiler/next/include/chpl/uast/Enum.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    return this->numChildren();
+  }
+  /**
+   Return the i'th EnumElement or Comment in this Enum.
+   */
+  const Expression* declOrComment(int i) const {
+    const ASTNode* ast = this->child(i);
+    assert(ast->isExpression());
+    return (const Expression*)ast;
+  }
```
Comment (lydia-duncan) [inline compiler/next/include/chpl/uast/Enum.h:81]: Nice!  These should help with that issue we had about commenting individual enum constants

File: compiler/next/test/frontend/testParseEnums.cpp [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+  auto a = enumDecl->declOrComment(0)->toEnumElementDecl();
+  auto b = enumDecl->declOrComment(1)->toEnumElementDecl();
+  assert(enumDecl->declOrComment(2)->isComment());
+  checkTest4(enumDecl, a, b);
+}
+
+static void test4j(Parser* parser) {
+  auto parseResult = parser->parseString("test4i.chpl",
+                                         "enum myEnum { a = ii, b, /*c*/ }\n");
+  assert(parseResult.errors.size() == 0);
```
Comment (lydia-duncan) [inline compiler/next/test/frontend/testParseEnums.cpp:290]: Shouldn't this be an error?

**Review by lydia-duncan — APPROVED** [review]
Just had one question, but otherwise I think this is good to go!

File: compiler/next/test/frontend/testParseEnums.cpp (hunk 3 above)
Comment (mppf) [reply compiler/next/test/frontend/testParseEnums.cpp:290]: Nope, we allow trailing commas in our enums, following C. It works in the regular compiler too.

**Pull request merged**